### PR TITLE
Enhancements to ObjectRegistry.

### PR DESCRIPTION
**Pull request by robertpustulka**
I've added `return $this` for method chaining and implemented `__set()`, `__isset()`, `IteratorAggregate` and `Countable`.

### CODE REVIEW

**Comment by markstory** [comment]
Reasonable thing to add, but you have failing tests, and need tests for the new methods.

**Review by markstory — CHANGES_REQUESTED** [review]
Reasonable thing to add, but you have failing tests, and need tests for the new methods.

**Comment by lorenzo** [comment]
Looks good to me :)

**Comment by robertpustulka** [comment]
@markstory is that OK to test new methods only in `ComponentRegistry`? `ObjectRegistry` is abstract and has never had it's tests.

**Comment by ADmad** [comment]
@robertpustulka I think having tests for `ComponentRegistry` is fine. If you were to add tests for the abstract class you would require phpunit mock which i believe internally would just make a class which extends the abstract class :)

**Comment by robertpustulka** [comment]
@ADmad I usually test abstract classes using mocks. This one doesn't have test, and I don't think it's necessary right now. Maybe in 4.0 :)

File: tests/TestCase/Controller/ComponentRegistryTest.php [hunk 1]
```diff
@@ -209,6 +209,22 @@ public function testUnload()
     }
 
     /**
+     * Test __unset.
+     *
+     * @return void
+     */
+    public function testUset()
```
Comment (markstory) [inline tests/TestCase/Controller/ComponentRegistryTest.php]: Typo in the method name.

**Review by markstory — APPROVED** [review]

File: tests/TestCase/Controller/ComponentRegistryTest.php (hunk 1 above)
Comment (robertpustulka) [inline tests/TestCase/Controller/ComponentRegistryTest.php]: @markstory I think you're referencing the old commit. It's been fixed.

**Comment by ADmad** [comment]
@robertpustulka Does this mean we can use `ObjectRegistry` in authentication plugin instead of custom collection class ? :)

**Comment by robertpustulka** [comment]
@ADmad we already do :)

**Pull request merged**